File: Niveau 4/Niveau 4.0/factorielles.hs
--------------------------------------------------------------------------------
-- Name:        Boîtes factorielles
-- Purpose:     France-IOI niveau 4.0
--
-- Author:      Sébastien
--
-- Created:     13/11/2022
--------------------------------------------------------------------------------

import Data.Array.Unboxed
import Data.List

maxNbCoeff = 12
factorielles :: Array Int Int
factorielles = a where a = array (1, maxNbCoeff) ((1, 1):[(i, i*a!(i-1)) | i <- [2..maxNbCoeff]])

main :: IO ()
main = do
    n <- readLn :: IO Int
    let decomposition = reverse . dropWhile (== 0) $ decompose n maxNbCoeff
    print $ length decomposition
    putStrLn $ intercalate " " . map show $ decomposition
    
decompose :: Int -> Int -> [Int]
decompose _ 0 = []
decompose n p =
    (n `div` factorielles!p):decompose (n `mod` factorielles!p) (p-1)
    

File: Niveau 4/Niveau 4.0/factorielles.py
#-------------------------------------------------------------------------------
# Name:        Boîtes factorielles
# Purpose:     France-IOI niveau 4.0
#
# Author:      Sébastien
#
# Created:     13/11/2022
#-------------------------------------------------------------------------------

MAX_NB_COEF = 13

def main():
    n = int(input())

    factorielle = [1]*MAX_NB_COEF
    for i in range(1, MAX_NB_COEF):
        factorielle[i] = i * factorielle[i-1]

    resultat = [0]*MAX_NB_COEF
    p = 0
    for i in range(MAX_NB_COEF-1, 0, -1):
        if n >= factorielle[i]:
            resultat[i] = n // factorielle[i]
            n %= factorielle[i]
            if p == 0: p = i

    print(p)
    print(" ".join(map(str, resultat[1:p+1])))

if __name__ == '__main__':
    main()
